sql: add description when removing FK while dropping a database

sajjadrizvi opened the pull request:
Previously, job descriptions were being populated improperly
for the GC jobs created through DROP DATABASE CASCADE. This
commit adds a meaningful fixes the description of the GC jobs,
which improves UX.

Release note: None

Fixes: #59221

alan-mas committed:
roachprod: refactor ssh.go
Third part of #47567 where we need to refactor ssh.go (as one type structure is in use, we cannot remove it completely).

So we are refactoring ssh.go and change its name to io.go and only keeping ProgressWriter struct and function

Also we are adding a unit test go code to ensure it works as expected.

Release note: None

Azhng committed:
sql: introduce crdb_internal.index_usage_stats virtual table

This commit introduce crdb_internal.index_usage_stats virtual
table that is backed by new clusterindexusagestats package. This
new package implements a variant of the indexusagestats interface
and serves the data by issuing cluster RPC fanout.

Addresses #64740

Followup to #66639

Release note (sql change): introduce crdb_internal.index_usage_statistics
virtual table to surface index usage statistics.
sql.metrics.index_usage_stats.enabled cluster setting can be used to
turn on/off the subsystem. It is default to true.

Azhng committed:
sql: implement persistedsqlstats flush logic

This commit implements the initial flush logic of the
persisted sql stats subsystem.

Release note: None

sumeerbhola committed:
admission,server: scope store overload admission control to write operations on the overloaded store

Previously KVWork was all admitted through a single WorkQueue, and this
WorkQueue was paired with a GrantCoordinator that took into account
both CPU overload and storage overload. This meant a single overloaded
store (in a multi-store setting) would slow down all KV operations
including reads, and those on stores that were fine.

We now have multiple WorkQueue's, one per store, that KV writes go
through, and then get to the shared WorkQueue that all operations go
through. The per-store WorkQueues are associated with their own
GrantCoordinators that listen to health state for their respective
stores. The per-store queues do not care about CPU and therefore do not
do grant chaining. Since admission through these per-store queues
happens before the shared queue, a store bottleneck will not cause KV
slots in the shared queue to be taken by requests that are still waiting
for admission elsewhere. The reverse can happen, and is considered
acceptable -- per-store tokens, when a store is overloaded, can be used
by requests that are now waiting for admission in the shared WorkQueue
because of a cpu bottleneck.

The code is significantly refactored for the above:
NewGrantCoordinators returns a container called StoreGrantCoordinators
which lazily initializes the relevant per-store GrantCoordinators when
it first fetches Pebble metrics, in addition to the shared
GrantCoordinator. The server code now integrates with both types and
the code in Node.Batch will sometimes subject a request to two
WorkQueues. The PebbleMetricsProvider now includes StoreIDs, and the
periodic ticking that fetches these metrics at 1min intervals, and does
1s ticks, is moved to StoreGrantCoordinators. This simplifies the
ioLoadListener which no longer does the periodic ticking and eliminates
a some testing-only abstractions. The per-store WorkQueues share the
same metrics, which represent an aggregate across these queues.

Informs #65957

Release note: None

Azhng committed:
sql: remove `count` from stmt/txn stats system table

Previously, system.statement_statistics and
system.transaction_statistics table includes a `count` column
that corresponds to `roachpb.StatementStatistics.Count` and
`roachpb.TransactionStatistics.Count` fields respectively.
The objective for that column is to make
`INSERT ON CONFLICT DO UPDATE` style query easy. However,
since early prototyping have shown that
`INSERT ON CONFLICT DO UPDATE` style statement is quite inefficient,
the SQL Stats flush mechanism will be implemented using
separate queries INSERT and UPDATE statements.
This column is no longer userful and it would require special handling.
Removing this column simplifies the flush logic and removes the
need for special handlings.

Release note (sql change): count column is removed from
 system.statement_statistics and system.transaction_statistics
 tables.

Commit:
jobs: create jobs in a batch

This commit adds a new function in Registry, 'CreateJobsWithTxn',
that allows creating jobs in a batch using a single SQL statement.
This function is a precursor to creating schema-changing jobs
in a batch, which will result in reducing the number of round-trips
between nodes during schema-changes.

Release note: None

dhartunian committed:
server: pod-to-pod fanout for statements api on tenant

Previously, the Statements endpoint implementation was purely local in
its execution and reported only in-memory data on the current tenant.
This change initializes a gRPC and gRPC-gateway server on the tenant,
much like we do on the hosts already, and uses the newly added
InstanceID subsystem for tenants to implement a gRPC-based fanout for
the Statements endpoint implementation.

The fanout is done much in the same manner as on hosts and we
expose the status server endpoints via HTTP on the tenant as well.

This change is necessary to support serverless observability features
and provide our UIs access to the Statements endpoint. Future work may
move this API to SQL-only

Resolves #64477

Release note (api change): tenant pods now expose the Statements API at
`/_status/statements` on their HTTP port.

pawalt committed:
docs: add ON UPDATE and row rehoming RFC

This PR adds an RFC for ON UPDATE constraints and
REGIONAL BY ROW rehoming.

Release note: None

adityamaru committed:
importccl: fail PGDUMP IMPORT if we see a UDT

In 21.1 we added `ignore_unsupported_statements` that ignored
all statements that we didnt support while parsing the PGDUMP
file. We don't support `CREATE TYPE` and its subsequent usage,
and this currently causes a NPE.

While we work on adding UDT support to PGDUMP we should make
an exception to the ignore option and fail the import instead.

This change also nils out the type resolver on the semaCtx that we
use during import table creation. This ensures that we do not
attempt to resolve UDTs but instead fail gracefully.

IMPORT TABLE and bundle imports do not support UDTs and so this
is an improvement over the NPEs import would hit prior to this patch.

Release note (bug fix): IMPORT PGDUMP with a UDT would result
in a nil pointer exception. This change makes it fail gracefully.

nvb committed:
kv: add active pushee variants TestLockTableWaiterWithErrorWaitPolicy

The test was not exercising the case where a pushee was still active
during the PUSH_TOUCH issued by requests with an Error wait policy.

nvb committed:
kv: add LockTimeout option to BatchRequest

Informs #67513.

This commit introduces a new `lock_timeout` field on the BatchRequest
Header struct.

lock_timeout specifies the maximum amount of time that the batch request
will wait while attempting to acquire a lock on a key or while blocking on
an existing lock in order to perform a non-locking read on a key. The time
limit applies separately to each lock acquisition attempt. If the timeout
elapses when waiting for a lock, a WriteIntentError will be returned.

Unlike in some other systems like PostgreSQL, where non-locking reads do
not wait on conflicting locks, in CockroachDB, non-locking reads do wait
for conflicting locks to be released. Because of this, the lock_timeout
configuration applies to non-locking reads in read-write and read-only
transactions as well.

Only the (default) Block wait policy will allow a request to wait on
conflicting locks, so the timeout only makes sense in conjunction with the
Block wait policy. The Error wait policy will throw an error immediately if
a conflicting lock held by an active transaction is encountered, so this
timeout can never be hit with an Error wait policy.

A value of zero disables the timeout.

Commit:
sql: create jobs in a batch in a transactions

This commit changes the way schema-change jobs are created in a
transaction. Previously, the jobs were created individually during the
execution of a transaction. A job creation incurs a round trip to
insert an entry in jobs table. This commit accumulates the jobs
to be created during a transaction and creates them in a batch, which
is expected to reduce the number of round-trips between distant nodes
during a transaction execution. This is a step forward to improve the
performance of GRANT/REVOKE queries in large databases.

Release note: None

Fixes: #64389

rafiss committed:
sql: settings from ALTER ROLE ... SET apply on session init

The default settings are applied when creating a new pgwire session. The
information is cached alongside authentication info to reduce the
latency of establishing a connection and minimize the chance of a login
availability hit.

Release note (sql change): Default session variable settings configured
by `ALTER ROLE ... SET` are now supported. The following order of
precedence is used for variable settings:

1. settings specified in the connection URL as a query parameter.
2. per-role and per-database setting configured by ALTER ROLE.
3. per-role and all-database setting configured by ALTER ROLE.
4. all-role and per-database setting configured by ALTER ROLE.
5. all-role and all-database setting configured by ALTER ROLE.

RESET does not validate the setting name.
SET validates both the name and the proposed default value.

Note that the default settings for a role are not inherited if one role
is a member of another role that has default settings. Also, the
defaults _only_ apply during session initialization. Using `SET
DATABASE` to change databases does not apply default settings for that
database.

The `public`, `admin`, and `root` roles cannot have default session
variables configured. The `root` role also will never use the "all-role"
default settings. This is so that `root` has fewer dependencies during
session initialization and to make it less likely for `root`
authentication to become unavailable during the loss of a node.

Changing the default settings for a role requires the role running the
ALTER command to either be an ADMIN or to have the CREATEROLE role
option. Only ADMINs can edit the default settings for another admin.
Futhermore, changing the default settings for `ALL` roles is
_only_ allowed for ADMINs. Roles without ADMIN or CREATEROLE _cannot_
change the default settings for themselves.

otan committed:
builtins: introduce 1 arg variant parse_interval / to_char

Release note (sql change): Introduce parse_interval and
to_char which takes in 1 string or interval and assumes the
Postgres IntervalStyle to make it's output.

otan committed:
builtins: add parse_date and to_char_with_style(date, ...)

Release note (sql change): Introduce a parse_date builtin with two
variants - the single arg variants parses a date and assumes ISO,MDY
datestyle and the two arg variants parses a date assuming the datestyle
variant on the second argument. This provides an immutable way of
casting strings to dates.

Release note (sql change): Introduce to_char(date), which
assumes a DateStyle of ISO,MDY and outputs date in that format. There is
also a to_char_with_style(date, DateStyle) variant which outputs the
date in the chosen DateStyle. This provides an immutable way of casting
dates to strings.

nvb committed:
kv: introduce request RoutingPolicy configuration

Touches #67562.

This commit introduces a new RoutingPolicy configuration that lives on a
BatchRequest header. A request's routing policy specifies how the
request should be routed to the replicas of its target range(s) by the
DistSender. There are initially two routing policies:
```
enum RoutingPolicy {
  // LEASEHOLDER means that the DistSender should route the request to the
  // leaseholder replica(s) of its target range(s).
  LEASEHOLDER = 0;
  // NEAREST means that the DistSender should route the request to the
  // nearest replica(s) of its target range(s).
  NEAREST = 1;
}
```

The default policy is `LEASEHOLDER`.

Routing policies allow us to stop overloading the use of the
ReadConsistency enum to dictate both how the client should route a
request to a server and which kinds of requests should be eligible to be
served by a given replica. Routing policies are a client-side only
concept. They do not dictate which replicas in a range are eligible to
serve the request, only which replicas are considered as targets by the
DistSender, and in which order. A request that is routed to an
ineligible replica (a function of request type, timestamp, and read
consistency) will be rejected by that replica and the DistSender will
target another replica in the range.

As discussed in https://github.com/cockroachdb/cockroach/pull/67725#pullrequestreview-711246613,
we will likely need to introduce a third routing policy that called
`SINGLE_REPLICA` to address #67554. This policy would be accompanied by
a ReplicaDescriptor and would specify that a given request must be sent
to that replica and DistSender should throw an error if the replica is
not part of its cached range descriptor. This is important to ensure
that a QueryResolvedTimestampRequest and its follow-up ScanRequest are
both sent to the same replica.

pawalt committed:
rfcs: respond to row rehoming rfc comments

Release note: None

stevendanna committed:
changefeedccl: allow changefeeds on regional by row tables

Previously we explicitly disallowed changefeeds on regional by row
tables because:

1) the Avro encoder lacked enum support,
2) the schema feed did not correctly identify RBR schema changes as
   both a column addition and a primary key change, and
3) we were unsure whether users would be OK seeing the new crdb_region
   column in their changefeed output.

Both (1) and (2) have since been fixed in
8113beda91a6c3791208632ecad075f6ca0a95b6 and
cd687ded12d717032f0966eca37b9c0845ee8d57, respectively.

For (3) we've decided that changefeeds should rather faithfully emit
what KV tells us about. Future work to support projections and
predicate filtering may allow _users_ to filter what gets output, but
until then we want to err on the side of avoiding ad-hoc modifications
to the changefeed output. Thus, we are OK emitting the crdb_region
column as-is both in the row data and as part of the primary key.

Release note (enterprise change): CHANGEFEEDs no longer fail when
started on REGIONAL BY ROW tables. Note that in REGION BY ROW tables,
the crdb_region column becomes part of the primary index. Thus,
changing an existing table to REGIONAL BY ROW will trigger a
changefeed backfill with new messages emitted using the new composite
primary key.

DrewKimball committed:
sql: add cluster setting to preserve subquery and cte ordering

This patch adds an experimental cluster setting `propagate_input_ordering`
that indicates whether subqueries and CTEs should propagate their orderings
to their parent scopes, when the parent scope is unordered. As an example,
the following two queries should produce the following result when the
cluster setting is true:
```
select * from (select * from generate_series(1, 10) i order by i % 5 asc, i asc) tmp;
with tmp as (select * from generate_series(1, 10) i order by i % 5 asc, i asc) select * from tmp;
----
   5
  10
   1
   6
   2
   7
   3
   8
   4
   9
```
This allows cockroach to imitate postgres behavior - while postgres
does not guarantee to maintain ordering on subqueries, it does in
practice. Some existing applications take advantage of this fact,
and so the ability to toggle this setting can help resolve
incompatibilities in some cases.

Fixes #68211

Release note: None

knz committed:
cli/zip: avoid quadratic behavior in SQL schema retrieval

Prior to this patch, the `debug zip` command would issue a
`DatabaseDetails` request over the network for each database in turn,
and also `TableDetails` for each table.

Server-side, each of these individual requests would cause all
descriptors to be loaded into memory first.

The overall behavior was thus quadratic in performance, and was
resulting in unacceptable delays on large clusters with thousands of
descriptors.

It turns out, none of this is necessary since we already collect the
descriptor details via `system.descriptors` and can collect the CREATE
statements trivially using the `crdb_internal.create_statements`
vtable, which also has linear performance.

Release note (cli change): `cockroach debug zip` no longer retrieves
database and table details into separate files. The schema information
is collected by means of `system.descriptors` and `crdb_internal.create_statements`.

rafiss committed:
sql/authentication: move GetDefaultSessionSettings to own function

Release note: None

rafiss committed:
sql: rename sql/authentication to sql/sessioninit

Add a few comments to increase readability.

Release note: None

otan committed:
sql: gate DateStyle behind datestyle_enabled session variable

Release note: None

knz committed:
clisqlshell: improve the behavior of `\c`

This patch addresses two shortcomings:

- `\c <dbname>` was previously implemented using SET using the same
  connection. However, as @rafiss points out, the team has been working
  on per-role default session variables, and those defaults are only
  applied when a new connection is created, and not when using SET. It
  would be strange for `\c` to block access to this functionality.

  To address this, `\c <dbname>` now uses the same reconnection logic.

- If the connection string was using TLS client certs, then changing
  the connection would always lose the TLS client cert details and force
  the user to enter a password (even when the account did not have a
  password).

  This patch fixes it.

Release note (cli change):  The `\c` client-side command in
`cockroach sql` and `cockroach demo` now always reconnects to the
server even when only changing the current database.
(This negates a part of a previous release note.)

knz committed:
clisqlshell: fix `\c` with insecure mode

The `\c` command was forgetting to disable TLS in the new URL if it
was disabled to start with. This fixes it.

Release note: None

jeffswenson committed:
sqlproxy: add a timeout when dialing tenant pods

Dialing dead or non existent pods on GKS can black hole
the network connection. This leaves the connection process
stuck until the OS TCP timeout of 2 minutes.

The SQL proxy and the tenant pod are colocated. 5 seconds should
be more than enough to establish the TCP connection. Timeing out
will cause the retry loop in proxy_handler.go to refetch the tenant
IP address and retry the connection.

Release note: None

otan committed:
builtins: implement parse_time, parse_timetz

Note to_char_with_style is not needed as these types do not need date to
output.

Release note (sql change): Implement parse_time and parse_timetz
builtins, which parses a time or timetz with immutable volatility.

Santamaura committed:
ui: wip add small component for ces survey link

Previously there was no ces survey link
It is important to get feedback from clients
Made a small component that will link to the ces survey

Release note (ui change): A ces survey link component was added
to support being able to get client feedback.

fqazi committed:
sql: remove table backrefs when dropping views in new schema changer

Previously, when dropping a view in the new schema changer
the table back refs weren't properly cleaned up. This was
inadequate because we left back references to views inside
table descriptors, which could lead to errors when dropping.
To address this, this patch adds logic to clean up view back
references to tables.

Release note: None

otan committed:
builtins: add timestamp/DateStyle variant for parse_timestamp/to_char

Release note (sql change): parse_timestamp now has a two argument
variant, which takes in a DateStyle and parses timestamps according to
that DateStyle. The one argument version assumes MDY. These builtins
have an immutable volatility.

Release note (sql change): Introduce a timestamp,DateStyle variant
to to_char_with_style, which converts timestamps to a string with
an immutable volatility. There is also a 1 arg to_char for timestamp
values which assumes the ISO,MDY output style.

otan committed:
sql: mark various datestyle casts as stable if DateStyle_enabled

Release note (sql change): In an earlier commit, CockroachDB now
uses the value of the `DateStyle` session var when interpreting
date to string conversions (and vice versa).

However, this made string::{date,timestamp} and
{date,timetz,time,timestamp}::string casts have a volatility of "stable"
instead of "immutable". This has ramifications for items such as
computed columns and check clauses, which cannot use immutable
expressions.

This means that the particular results returned by these queries can
become incoherent when `DateStyle` is customized to a different
value from its default, `ISO,MDY`.

In order to provide guardrails against this incoherence, CockroachDB
now provides a new, separate configuration knob called
`datestyle_enabled`, that applications can use to "opt in" the
ability to customize DateStyle:
- by default, this knob is false and applications cannot customize
  `DateStyle`. Invalid conversions that are already
  stored in the schema are unaffected and continue to produce
  results as per the default style `postgres`.
- when the knob is true, these things happen:
  - apps can start customizing `DateStyle` in SQL sessions.
  - a SQL session that has a custom `DateStyle` will start
    observing *incoherent results* when accessing tables that
    already contain the aforementioned casts.
  - new schemas cannot have the above casts.

The knob works as follows: the *primary* configuration mechanism
is a new cluster setting called `sql.defaults.datestyle.enabled`.
This is the knob that operators and DBAs should customize manually.
Then as a *secondary* configuration mechanism, the value of the
cluster setting is also copied to each SQL session as a new session
var `datestyle_enabled`. This is a performance optimization.
SQL apps should not modify this session var directly, except
for temporary testing purposes.

In v22.1, upgrades will be disabled if these stable expressions are
found in computed columns, check clauses, etc.

stevendanna committed:
storage: reduce allocations in MVCCIncrementalIterator

This changes some uses of Key() to UnsafeKey() to avoid unnecessary
allocations.  I've added a small benchmark that shows the decreased
number of allocations when iterating.

```
name                                            old time/op    new time/op    delta
MVCCIncrementalIterator/useTBI=true/ts=5-16       12.5ms ± 7%     9.0ms ± 7%  -28.02%  (p=0.000 n=57+26)
MVCCIncrementalIterator/useTBI=true/ts=480-16     2.36ms ± 8%    2.27ms ±13%   -4.12%  (p=0.000 n=52+25)
MVCCIncrementalIterator/useTBI=false/ts=5-16      6.32ms ± 7%    6.44ms ±11%     ~     (p=0.083 n=55+24)
MVCCIncrementalIterator/useTBI=false/ts=480-16    1.14ms ±10%    1.15ms ± 6%     ~     (p=0.390 n=54+25)

name                                            old alloc/op   new alloc/op   delta
MVCCIncrementalIterator/useTBI=true/ts=5-16        831kB ± 0%      32kB ± 0%  -96.20%  (p=0.000 n=57+23)
MVCCIncrementalIterator/useTBI=true/ts=480-16     60.5kB ± 0%    35.7kB ± 0%  -41.00%  (p=0.000 n=56+26)
MVCCIncrementalIterator/useTBI=false/ts=5-16      17.7kB ± 0%    17.7kB ± 0%     ~     (p=0.127 n=50+25)
MVCCIncrementalIterator/useTBI=false/ts=480-16    21.8kB ± 0%    21.8kB ± 0%     ~     (p=0.469 n=57+26)

name                                            old allocs/op  new allocs/op  delta
MVCCIncrementalIterator/useTBI=true/ts=5-16         103k ± 0%        0k ± 0%  -99.71%  (p=0.000 n=57+26)
MVCCIncrementalIterator/useTBI=true/ts=480-16      3.76k ± 0%     0.55k ± 0%  -85.28%  (p=0.000 n=57+26)
MVCCIncrementalIterator/useTBI=false/ts=5-16        22.0 ± 0%      22.0 ± 0%     ~     (all equal)
MVCCIncrementalIterator/useTBI=false/ts=480-16       277 ± 0%       277 ± 0%     ~     (all equal)
```

The benchmark uses data that I believe fits entirely into the cache,
which may explain why the TBI optimization is slower even when using a
timestamp that should exclude a large portion of the data.

Release note: None

nvb committed:
kv: assert txn unused in SetFixedTimestamp

This commit asserts that a transaction has not been used to read or to
write by the time that `SetFixedTimestamp` is called on it.

This was extracted from #68194 and modified to return an error from
`SetFixedTimestamp` on misuse instead of fatal-ing. This provides a
sufficient, temporary backstop for #68216 until the conn executor logic
is fixed:

```
root@127.0.0.1:26257/movr> create table t (x int);
CREATE TABLE

root@127.0.0.1:26257/movr> insert into t values (1);
INSERT 1

root@127.0.0.1:26257/movr> select crdb_internal_mvcc_timestamp, * from t;
   crdb_internal_mvcc_timestamp  | x
---------------------------------+----
  1628094563935439000.0000000000 | 1
(1 row)

root@127.0.0.1:26257/movr> begin as of system time (1628094563935439000.0000000000-1)::string;
BEGIN

root@127.0.0.1:26257/movr  OPEN> select * from t;
  x
-----
(0 rows)

root@127.0.0.1:26257/movr  OPEN> prepare y as select * from t as of system time 1628094563935439000.0000000000;
ERROR: internal error: cannot set fixed timestamp, txn "sql txn" meta={id=e5e81c19 pri=0.01517572 epo=0 ts=1628094563.935438999,0 min=1628094563.935438999,0 seq=0} lock=false stat=PENDING rts=1628094563.935438999,0 wto=false gul=1628094563.935438999,0 already performed reads
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:1016: SetFixedTimestamp()
github.com/cockroachdb/cockroach/pkg/kv/txn.go:1200: SetFixedTimestamp()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:278: populatePrepared()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:220: func1()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:226: prepare()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:112: addPreparedStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:570: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:126: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1626: func1()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1628: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1550: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:627: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:645: func1()
runtime/asm_amd64.s:1371: goexit()

HINT: You have encountered an unexpected error.

Please check the public issue tracker to check whether this problem is
already tracked. If you cannot find it there, please report the error
with details by creating a new issue.

If you would rather not post publicly, please contact us directly
using the support form.

We appreciate your feedback.

root@127.0.0.1:26257/? ERROR>
```

Santamaura committed:
ui: Complete implementation of survey link

There was no link to the feedback survey
Survey feedback is useful for improving the product
Added a small component to open the survey

Release note (ui change): added component to support survey link

darinpp committed:
sql/tenant: remove idle detection and exit flag

We previoulsy had a flag `--idle-exit-after` that was used to monitor the
active SQL connections and terminate the server if there are no more
clients. We don't need that functionality anymore so this PR removes it.
It also fixes the issues with the related flaky tests.
Fixes #66767
Release note: None

rickystewart committed:
ci: separate out "check generated files up-to-date" into separate ci job

This splits the lint build configuration roughly in half. The lint build
config is one of the longer-running ones we see in CI, so this should
save quite a lot of time.

Release note: None

RaduBerinde committed:
multitenant: report SQL pod CPU usage

The tenant cost controller now reports the user CPU time.

We have a built-in allowance of `1%` of a CPU core, to account for
background activity; only usage above that counts toward consumption.

Release note: None

nvb committed:
kv: include RangeID in rangefeed goroutine stacks

This commit includes the RangeID in each of a rangefeed processor and
its registations' associated goroutine stacks. This is a cheap and easy
way to get better observability into the ranges that have active
rangefeeds. It also tells us where those goroutines are spending their
time.

This will also become easier to use in Go 1.17, which improved the
format of stack traces.

rafiss committed:
parser: add VIRTUAL syntax to help

Release note: None

RaduBerinde committed:
opt: support enum types in test catalog

Release note: None

Commit:
sql: add description when removing FK while dropping a database

Previously, job descriptions were being populated improperly
for the GC jobs created through DROP DATABASE CASCADE. This
commit adds a meaningful fixes the description of the GC jobs,
which improves UX.

Release note: None

Fixes: #59221

Commit:
sql: add description for jobs created during DROP CASCADE

Previously, job descriptions were empty for GC jobs that
are created during DROP DATABASE CASCADE. This commit adds
descriptions of jobs.

Release note: None

Fixes: #59221

rickystewart committed:
ci: add bazel lint job for ci

Closes #67138.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68453)
<!-- Reviewable:end -->

nvb committed:
kv/rangefeed: avoid race between new registration and processor shutdown

This commit resolves a race between rangefeed registration connection
and disconnection which could have resulted in a leaked goroutine. The
call to `registration.runOutputLoop` is done in an async task, so it is
not synchronized with the calls to `registration.disconnect` from the
rangefeed processor (from `DisconnectWithErr`). Because these calls were
not synchronized, it was possible for `disconnect` to see a nil
`outputLoopCancelFn` and then shut down the stream. Then `runOutputLoop`
could set `outputLoopCancelFn` and enter `registration.outputLoop`. I
believe this could allow the output loop goroutine to leak, especially
if it never sent anything on the stream.

I don't think this relates to the resolved timestamp stall we're
currently investigating because `registration.disconnect` would have
still returned and error to the rangefeed stream during this race, so it
would not have resulted in a rangefeed getting stalled indefinitely.
Instead, only the outputLoop goroutine could have stalled, but we saw no
indication of this in goroutine dumps.

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sajjadrizvi)

___
*[pkg/sql/drop_cascade.go, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68453#-MgIITlhB2WMrGivHbcX:-MgIITlhB2WMrGivHbcY:bi2lpza) ([raw file](https://github.com/cockroachdb/cockroach/blob/9806d6685c57a6674108274b63a74e334872929d/pkg/sql/drop_cascade.go#L198)):*
> ```Go
> 		//TODO (sajjad): Add meaningful job descriptions instead of empty strings.
> 		if desc.IsView() {
> 			cascadedObjects, err = p.dropViewImpl(ctx, desc, false /* queueJob */, "", tree.DropCascade)
> ```

I am leaning towards adding `p.stmt` as the description to begin with.


<!-- Sent from Reviewable.io -->

RichardJCai committed:
sql: refactor default privileges for all roles

Default privileges for all roles are refactored to be
more in line with regular roles.

Simplifies some code.

Release note: None

RaduBerinde committed:
sql: add missing overload for enum[] IS NULL

This change adds an overload for AnyArray IS NULL. This covers cases
where the array element type is an enum.

The expression passes initial type-checking because it is parsed into
a special `IsNotNullExpr`, not a comparison. But the optimizer treats
it as a comparison and needs to find an overload.

Fixes #68233.

Release note (bug fix): fixed internal error in IS (NOT) NULL
expressions where the left operand is an array with a custom element
type (e.g. enum).

dt committed:
backupccl: use separate BACKUP and Export target file sizes

The BACKUP processor should pick its target file size -- at which it closes
a file and opens a new one -- independently from the KV's target Export file
size, which has factors like request duration, memory footprint, etc that may
influence it but are not relevant to the backup file as it is streamed our to
cloud storage.

Release note: none.

dt committed:
backup: always write to cloud storage in processor

This removes the bulkio.backup.proxy_file_writes.enabled setting, and just
always does the behavior of it being true.

Release note (sql change): the setting bulkio.backup.proxy_file_writes.enabled is no longer needed to enable proxied writes which are now the default.

vladlos committed:
ui: tooltip adjjustments

there were 2 issues with tooltiops on cluster-ui
pages:
 - stuck visible tooltip while scrolling
 - click on table header tooltip trigerred sort
fixed both in new ui-component version

Resolves: #67767, #67721

Release note(ui): fix tooltips behavier on session,
statements and trasaction pages.

dt committed:
backup: reduce default kv.bulk_sst.target_size to 16mb

Release note: none.

ajwerner committed:
sqlliveness/slstorage: record and log when deleting a session

There are two paths which delete sessions: a background job and on demand
when there's a cache miss. The second one, added later, and more commonly hit
did not log or record metrics.

Release note (ops change): Record a log event and counter increment when
removing an expired session.

DuskEagle committed:
release: update cockroach-cloud-images repository location

This commit updates the location of the cockroach-cloud-images Docker
repository. For reasons described in https://cockroachlabs.atlassian.net/browse/CC-4667,
we are switching our internal CockroachDB Docker repository to this location,
which uses Google Artifact Registry.

Release note: None

dt committed:
backup: add some VEventfs to backup file flushing

Release note: none.

dt committed:
backup: fix bug when counting progress for returned SSTs

Release note: none.

RaduBerinde committed:
sql: fix accidental FK check skips

This change fixes a bug in the insert fast path where we accidentally
skip subsequent FK checks when a FK check can be skipped due to NULL
value.

Note that this bug does not manifest when optbuilder can determine
that the check can be elided entirely; notably, this is always the
case for non-prepared statements which insert a single row.

Fixes #68307.

Release note (bug fix): fixed missing foreign key checks in some cases
when there are multiple checks and the inserted data contains a NULL
for one of the checks.

rickystewart committed:
ci: mount git alternates directory in container

We already do this in `build/builder.sh`.

Release note: None

Santamaura committed:
Use short circuit evaluation in render

HonoreDB committed:
kv,changefeedccl: do not fatal or panic on unrecognized retry errors

Change a panic and a log.Fatalf to just return an AssertionFailed error
instead, which should be less catastrophic in a mixed-version situation.

Honestly we could probably just default to ignoring these entirely
but that's potentially an infinite retry bug.

Closes #68469

Release note: None

nvb committed:
kv: don't load applied index twice during Raft snapshot

This was harmless, but it was also unnecessary and looked quite
alarming. It would be a major problem if the Raft applied index stored
in a `SnapshotMetadata` was ever incoherent with the applied state in
the snapshot. However, we're reading from a consistent enging snapshot
in `snapshot`, so the Raft applied index can't change. Still, better to
avoid questions.

The "members of the Range struct" comment was very stale, dating back
to acc41dd.

HonoreDB committed:
kv: version gate new error types to avoid fatal in the client
Depends on #68495
Fixes #68465

Release note: None

taroface committed:
util/log,docs/generated: update crdb-v2 example formatting

rickystewart committed:
dev: teach dev how to lint

Closes #68547.

Release note: None

catj-cockroach committed:
security: fixes CertificateLoader tests when running as root.

This change fixes the CertificateLoader tests when running as the
root user, since the acceptable permissions change depending on
what user owns the certificate files (which is root since the tests
create the certificate files).

Release note: None

HonoreDB committed:
changefeedccl: remove fatals

All events we get from a kvfeed are currently one of two types,
so in a bunch of places we call log.Fatalf if we can't categorize
an event into one of those two. This kills the entire server, which
seems like an overreaction. Let's preemptively change this to a
normal error so that new event types can be more safely added in the
future.

Release note: None

rickystewart committed:
dev: rip recording logic out of `dev` datadriven tests

This feature isn't used by anyone that currently works on `dev`, and
manually updating the recording files seems to be the most convenient
way to work on these tests.

In the future we may consider migrating to a proper mocking framework
for `dev`, but for now this structure is fine.

Also rename recorder -> recording for accuracy.

Release note: None

dt committed:
backup: use processor's locality to pick backup partition

Locality-partitioned backups were intended to reduce cross-regional file writes.
Thus, if the processor is writing a file, it should do so to the region that matches
its locality. Due to PartitionSpans, this should generally be the same node and thus
same region as the leaseholder, or at least where the leaseholder was during planning.

Release note: none.

yuzefovich committed:
sql: change physical planning heuristics a bit to prefer local execution

This commit changes two parts of the physical planner heuristics:
- we now say that the lookup join "can be distributed" rather than
  "should be distributed"
- for top K sort we also say that it "can be" rather than "should be"
  distributed.

I'm not certain whether both of these changes are always beneficial, but
here is some justification.

The change to the lookup join heuristic will make it so that the
distribution of the join reader stage is determined by other stages of
the physical plan in `distsql=auto` mode. Consider an example when the
input to the lookup join is the table reader that scans only a handful
of rows. Previously, because of the "should distribute" heuristic, such
a plan would be "distributed" meaning we would plan a single table
reader on the leaseholder for the relevant range (most likely a remote
node from the perspective of the gateway node for the query); this, in
turn, would force the planning of the join reader on the same node, and
all consequent stages - if any - too. Such a decision can create
a hotspot if that particular range is queried often (think append-only
access pattern where the latest data is accessed most frequently).

With this change in such a scenario we will get more even compute
utilization across the cluster because the flow will be fully planned on
the gateway (which assumed to be chosen randomly by a load balancer),
and the lookup join will be performed from the gateway (we'll still need
to perform a remote read from the leaseholder of that single range).

The change to the top K sort heuristic seems less controversial to me,
yet I don't have a good justification. My feeling is that usually the
value of K is small, so it's ok if we don't "force" ourselves to
distribute the sort operation if the physical plan otherwise isn't
calling for it.

Overall, the choice of making changes to these two heuristics isn't very
principled and is driven by a single query from one of our largest
customers which happened to hit the hot spot scenario as described
above. In their case, they have append-like workload that is constantly
updating a single range. Eventually that range is split automatically,
but both new ranges stay on the same node. The latest data is accessed
far more frequently than any other data in the table, yet according to
the KV heuristics the ranges aren't being reallocated because the scans
hitting the hot ranges aren't exceeding the threshold. What isn't
accounted for is the fact that other parts of the flow are far more
compute-intensive, so this change attempts to alleviate such a hot node
scenario.

Release note (sql change): Some queries with lookup joins and/or top
K sorts are now more likely to be executed in "local" manner with
`distsql=auto` session variable.

rafiss committed:
*: update to lib/pq v1.10.2

This gets us off the cockroachdb/pq fork.

This revealed a bug in some of the tests: they were not calling
rows.Close() after querying the database. The new version of
lib/pq is more strict about requiring this.

Additionally, logic tests were not checking `NextResultSet()`, which is
a bug if the test executes multiple queries at once.

Release note: none

stevendanna committed:
roachtest,workload,roachprod: add version command and --version flag

It is often important to capture the versions of these tools used
during a test.  To help facilitate this, each command now supports a
`version` command that outputs build information captured at link
time.

This information is also available via the `build` package, which will
be useful for outputting it in other forms in the future as well.

Release note: None

sumeerbhola committed:
tests: add tpccbench,kv0,kv95 with admission control enabled

See https://github.com/cockroachdb/cockroach/issues/65955#issuecomment-893645465
for some observations from manual runs of these.

Release note: None

dt committed:
backup: log warning on unexpected multi-file response

With header.TargetBytes = 1 we expect each request to return at most one file,
but the loop, kept here for legacy reasons may suggest otherwise. Logging a warning
should help document what we expect while also informing us if we're mistaken.

Release note: none.

Commit:
backupccl: add option to display descriptor ids to SHOW BACKUP

Previously there weren't any descriptor ID information displayed by
SHOW BACKUP. These IDs provide useful information during triage and
general debugging. To address this, a WITH option debug_ids was added
to populate additional columns in SHOW BACKUP with descriptor IDS of
all objects along with the IDs of its database and parent schema.

Release note (enterprise change): Descriptor IDs of every object are now visible in SHOW BACKUP,
along with the descriptor IDs of the object's database and parent schema. SHOW BACKUP will display
these IDs if the `WITH debug_ids` option is specified.

rickystewart committed:
ci: split maybe_stress, maybe_stressrace into their own build configs

Closes #68380
Closes #68381

Release note: None

rail committed:
acceptance: explicitly specify binary path

Touches #59446

Previously, in order to run the acceptance test we were implicitly
identifying the `cockroach` binary to the acceptance test. This is a bit
problematic under bazel, where the binary location may be in different
places.

This patch explicitly passes the location if the cockroach binary to the
acceptance test.

Release note: None

rafiss committed:
sql: fix COPY CSV so it handles multiple records at a time

This was broken since the golang csv reader reads the entire input all
at once, so the underlying buffer is consumed. This caused the loop that
reads each record to terminate early.

Release note (bug fix): Fixed the COPY CSV command so that it handles
multiple records separated by newline characters.

knz committed:
sql: regression test for pg's INET format

PostgreSQL and CockroachDB recognize `10.0.0.017` as `10.0.0.17` when
interpreted using SQL type INET.

However, Go 1.17's IP address parser will soon change to prevent
this. We do not want a change in Go to surrepticiously change crdb's
behavior. So this commit adds a test to validate the behavior.

Release note: None

adityamaru committed:
go.mod: bump grpc to v1.39.1

Release note: None

otan committed:
roachtest/pgjdbc: enable DateStyle test

Release note: None

fqazi committed:
sql: drop view cascade does not handle deps being dropped earlier

Fixes: #68600

Previously, drop view cascade would only check if a dependency
was dropped when initially going over the list. It did not check
before executing the drop logic, so an earlier dependency could
cause the current object to move to a dropped state. This was
inadequate because cascaded drops of views could fail due to this
reason. To address this, this patch skips dropping dependencies
if at a later stage they are found to be dropped.

Release note (bug fix): Cascaded drop of views could run into
'table ...is already being dropped' errors incorrectly.

ajwerner committed:
sql: rework SHOW JOBS WHEN COMPLETE delegate

The previous pattern was fragile because the filter could be re-arranged.

Release note: None

mgartner committed:
cloud: bump orchestrator version to 21.1.7

Release note: None

mgartner committed:
roachtest: bump predecessor to v21.1.7

Release note: None
